Fix inflate() flush modes and implement state.data_type

cjgriscom opened the pull request:
This PR address the following issues:

- `state.data_type` is always 0
- https://github.com/memorysafety/zlib-rs/issues/114
- Add logic to advance from `TYPE` state to `TYPE_DO` state

folkertdev commented on zlib-rs/src/inflate.rs:
```diff
@@ -1922,32 +1927,38 @@ pub unsafe fn inflate(stream: &mut InflateStream, flush: InflateFlush) -> Return
         stream.msg = msg.as_ptr() as *mut u8 as *mut core::ffi::c_char;
     }
 
+    stream.data_type =
+        state.bit_reader.bits_in_buffer() as i32 +
+        if state.last { 64 } else { 0 } +
+        if matches!(state.mode, Mode::Type) { 128 } else {0} +
+        if matches!(state.mode, Mode::Len | Mode::CopyBlock) { 256 } else { 0 };
+
```
because we're only ever in one mode, I think a `match` is nicer here. Also can tests be added that verify the behavior (probably from the libz-rs-sys` crate's test modules) versus zlib itself?

cjgriscom replied:
Hmm, you're right.  I will double-check zlib's original logic and correct it to one match.  I will also put some tests together that demonstrate parity b/t libz-rs-sys and libz-sys

folkertdev commented on zlib-rs/src/inflate.rs:
```diff
@@ -1880,6 +1880,11 @@ pub unsafe fn inflate(stream: &mut InflateStream, flush: InflateFlush) -> Return
     let dest_slice = core::slice::from_raw_parts_mut(stream.next_out, stream.avail_out as usize);
 
     let state = &mut stream.state;
+    if matches!(state.mode, Mode::Type) {
+        state.mode = Mode::TypeDo;
+    }
```
can a test be constructed that verifies this behavior?

folkertdev reviewed (commented):
nice work! 

it's interesting that none of the zlib/zlib-ng tests catch these bugs. We'd like to add additional tests when we find that incorrect behavior slipped through.

cjgriscom commented:
I split the changes into 3 commits and added a test which compares the stream positions and state data after each call to inflate().  It demonstrates that all three changes are needed for consistency

folkertdev commented:
After your changes, other tests were failing. Turns out this is what the `LEN_` state is for. It looks really weird otherwise because it does no real work, just falls throught, but apparently that state is important for the correct behavior of `Flush::Trees`.

Also for the future, our CI checks formatting and runs `cargo clippy` (also on tests). We advise you run those tools locally at every commit, that'll save a bunch of back-and-forth with CI.

folkertdev reviewed (approved):
Thanks for these fixes!

cjgriscom commented:
Thanks for merging and for your work on this crate!
